### ISSUE TITLE
Bad accessbillity

### ISSUE THREAD

**Issue by anotherdoesnm** [trufae/parla#57]
Hello, the app is cool but can't be used by my friend who is completely blind because screenreader doesn't see the window and seems it's hard to use it without mouse

**Comment by trufae** [comment]
the app can be used without a mouse, i use it with the keyboard all the time, but maybe i'm missing some interactions. i never used a screenreader but i would appreciate some guidance to reproduce or understand better how to reproduce the problem.

**Comment by trufae** [comment]
hopefully addressed in https://github.com/trufae/parla/pull/65

**Comment by pvagner** [comment]
Hello,

Unfortunatelly there are more screen reader accessibility issues even on linux.

These are the ones I can recognize. Let me know if you would like to describe them in more details or filled as separate issues please.

* Problematic application wide ctrl+tab / ctrl+shift+tab keyboard shortcuts. With GTK4 this is used to navigate out of list. For example when focus is in the list of conversations pressing ctrl+shift+tab should move the focus back to the *Search contacts* text entry.
* When navigating in the list of conversations conversation is switched into immediately when changing selection. This is problematic since in order to switch into a conversation I will inadvertently mark all the conversations passing by as read. It should only switch when pressing the enter key or double clicking.
* Inability to inwoke the popup menus from the keyboard, filled already as #40
* While typing a message the conversation list item is being updated with the content of the message draft and it's constantly being reported with a screen reader leading to unwanted over-verbosity.
* When browsing the list of messages in a conversation, labels with the message content are focusable so text selection is possible. However individual list items are not focusable thus it's difficult to find out message properties such as sender, date, status etc. Also having the list items focusable is a prerequisite for being able to inwoke the message popup menu. Also ctrl+tab / ctrl+shift+tab is also blocked in this part of the UI.

Let me know if you need more explanations or additional steps for reproducing these.

Greetings

Peter

**Comment by trufae** [comment]
Hopefully improved in 0.8.8

**Comment by pvagner** [comment]
Hello @trufae,

The 0.8.8 update is massive in terms of a11y.

* Message reading is awesome now, its popup menu as well. If you would like to polish it even further there's another enhancement request from me:
    - After deleting a message or receiving a message or other updates , message list focus falls back to the first message in the list. It would be helpfull to preserve the focus on update or move focus to the previous message to the one that has just been deleted if it's doable.
* All the message types including calls, missed calls, location streaming, info messages, all of them are clearly reported and can be navigated to with a screen reader. Huge compliments for implementing it like this!
* And I have a question related to the message list: does it support selecting multiple messages? I think not, I just would like to make sure I am understanding it right.
* Sidebar conversation list is now great too. Navigation is working fine with arrow keys, page up / page down as well as home and end keys according to the GTK convention. There are some more enhancement requests related to the sidebar conversations list:
    - When the sidebar is collabsed by pressing ctrl+shift+s it displays conversation avatars and unread badges. Although the avatar images inside are labelled the focused list items have no accessible name when the sidebar is in this state. It would be helpfull to set it explicitly and also include the unread messages counter.
    - I think conversations list has a popup menu as well and it's not yet accessible by pressing shift+F10 and the applications key. Can it be added please?
    - When pressing ctrl+tab while the conversations list is focused you have designed it so the chat entry receives the focus. My thinking is that it would be better to preserve the natural focus order here by focusing the first element inside the toolbar with buttons such as *Hide sidebar*, *Search in conversation* and so on. Is that doable and do you also like it?
    - When pressing ctrl+s keyboard shortcut to Show and Hide the sidebar, would you consider also manipulating the focus? When the state changes from hidden to showing it would be nice to focus the item corresponding to the active chat. When changing state from showing to hidden it would be nice to place the focus either to the chat entry or first item on the toolbar (currently show sidebar button). This is not accessibility as such as this part is already accessible I am seeing this as an efficiency and general usability improvement request. If you don't like, feel free to ignore this suggestion.
* When playing voice messages there is a shared media player with its controls. There is a playback speed combobox. When the combobox is expanded the first item in the list is preselected. Is that under your control so you can make the active value preselected instead? Again this part is accessible already, this is just another convenience enhancement request.
* When using alt+tab to navigate over all open apps on the gnome desktop, parla's icon is not labelled, it's reported like io.github.trufae.parla instead. I don't know how to improve this but it would be nice to setup an accessible name *Parla* somewhere. I can recognize the application window is already called *Parla*.
* And finally thank you for the nice accessible media viewer. The list of audio files is great, its popup menu is accessible from the keyboard, and there is super nice unique feature that allows saving all the media files from a conversation into a folder. This save folder thingy is not screen reader accessibility related but I do like it very much so I couldn't resist.

Once again huge huge thanks, these are excelent improvements!

**Comment by SeanRandall** [comment]
I'm running [NVDA](https://nvaccess.org) on Windows, although for testing windows Narrator will give you similar results.  My worst accessibility gripe so far in 0.8.8 is the list of conversations in the sidebar don't announce when you cursor through them. NVDA sees the items as list items, but their text appears empty.

@pvagner  told me about ctrl+k which works better, so if you could replicate the functionality of that list somehow for screen reader access in the sidebar it'd be very helpful!
I didn't see any earlier versions but this is usable, if not perfect. :) Thanks for all your efforts!

Oh, and the main menu doesn't read either, I have to use OCR on the screen to click items in there. popup menu, such as on a message, works fine.

**Comment by trufae** [comment]
try now in 0.9.0 :3

**Comment by pvagner** [comment]
Hello @trufae 

Almost all the things I've asked for are now solved.

* [ ] When using alt+tab for switching apps the screen reader is still reporting *io.github.trufae.parla* and I can't figure it out either.
* [ ] Now expanded and collabsed sidebar conversations list rows read the same way with a screen reader no matter the state. Perhaps you should consider leaving out out the message preview if it's not visible. It's a very minor thing in a way to full ux polishing as the message preview is spoken last with a screen reader so we can skip it easily.
* [ ] I have seen you mentioning message selection checkbox in 2c03c6134e245486027401f9ed66a481bf806227. But still I don't know when it's displayed.
* [ ] When navigating in the list of messages I can't get return to reply working. If I press the Reply button in a popup menu the reply is working fine.
* [x] Next to the chat input text entry there is a label saying "Type a message". Can this be updated to say "Type a reply" when replying and then setup as a labelled by relation for the chat input text entry please? It would be reported along side the text entry when focused and we would be able to find out easily if we are replying or sending a new message.

It's all amazing on linux now. Let's wait for Windows guys to chime in if there's something to improve.

Oh btw in this version default gstreamer based audio playback has started working for me on linux including playback speed. Thank you!

**Comment by pvagner** [comment]
Dear @trufae 
Huge thanks for even more screen reader accessibility improvements. Excuse me for being too picky but I'm still wondering about a few things:

* Related to #81 it's working great for the time being, however wouldn't you like to make it so pressing the enter key also respects double click action configured in the settings, such as reacting with thumbs up or viewing contact details?
* I still can't figure out how selecting multiple messages is supposed to work. When using up and down arrow keys to navigate in the list of messages in a conversation, how do I start selecting messages? Can you please try explaining it as a list of steps mentioning the keys to press please?

Huge thanks for another awesome update.
